cockatrice/src/spoilerbackgroundupdater.cpp

```diff
@@ -0,0 +1,210 @@
+{
+    // Check for server reply
+    auto *reply = dynamic_cast<QNetworkReply *>(sender());
+    QNetworkReply::NetworkError errorCode = reply->error();
+
+    if (errorCode == QNetworkReply::NoError)
+    {
+        spoilerData = reply->readAll();
+        reply->deleteLater();
+
+        // Save the spoiler.xml file to the disk
+        saveDownloadedFile(spoilerData);
+    }
+    else
+    {
+        qDebug() << "Error downloading spoilers file" << errorCode;
+    }
+}
+
+void SpoilerBackgroundUpdater::actCheckIfSpoilerSeasonEnabled()
+{
+    auto *response = dynamic_cast<QNetworkReply *>(sender());
+    QNetworkReply::NetworkError errorCode = response->error();
+
+    if (errorCode == QNetworkReply::ContentNotFoundError)
+    {
+        // Spoiler season is offline at this point, so the spoiler.xml file can be safely deleted
+        // The user should run Oracle to get the latest card information
+        QString fileName = settingsCache->getSpoilerCardDatabasePath();
+        QFileInfo fi(fileName);
+        QDir fileDir(fi.path());
+        QFile file(fileName);
+
+        // Delete the spoiler.xml file as we're not in spoiler season
+        if (file.exists() && file.remove())
+        {
+            qDebug() << "Spoiler Season Offline, Deleting spoiler.xml";
+        }
+
+        /*
+         * ALERT: Ensure two reloads of the card database do not happen
+         * at the same time or a racetime condition can/will happen!
+         */
+        qDebug() << "Spoiler Season Offline, Reloading Database";
+        QtConcurrent::run(db, &CardDatabase::loadCardDatabases);
+    }
+    else if (errorCode == QNetworkReply::NoError)
+    {
+        qDebug() << "Spoiler Service Online";
+        startSpoilerDownloadProcess(SPOILERS_URL, true);
+    }
+    else
+    {
+        qDebug() << "Error: Spoiler download failed with reason" << errorCode;
+    }
+}
+
+bool SpoilerBackgroundUpdater::saveDownloadedFile(QByteArray data)
+{
+    QString fileName = settingsCache->getSpoilerCardDatabasePath();
+    QFileInfo fi(fileName);
+    QDir fileDir(fi.path());
+
+    if (!fileDir.exists() && !fileDir.mkpath(fileDir.absolutePath()))
+    {
+        return false;
+    }
+
+    // Check if the data matches. If it does, then spoilers are up to date.
+    if (getHash(fileName) == getHash(data))
+    {
+        if (trayIcon)
+        {
+            trayIcon->showMessage(tr("Spoilers already up to date"), tr("No new spoilers added"));
+        }
+        qDebug() << "Spoilers Up to Date, Reloading Database";
+        QtConcurrent::run(db, &CardDatabase::loadCardDatabases);
+        return false;
+    }
+
+    QFile file(fileName);
+    if (!file.open(QIODevice::WriteOnly))
+    {
+        qDebug() << "Spoiler Service Error: File open (w) failed for" << fileName;
+        return false;
+    }
+
+    if (file.write(data) == -1)
+    {
+        qDebug() << "Spoiler Service Error: File write (w) failed for" << fileName;
+        return false;
```

Comment:
The file handle won't be closed in this branch I think?

Reply:
Never has been, but I can make it close